File: 정선웅/소수찾기.java
import java.util.*;

class Solution {
    Set<Integer> primeCandidates = new HashSet<>();
    boolean[] visited;

    public boolean isPrime(int num) {
        if(num < 2)
            return false;

        for(int i=2; i*i<=num; i++) {
            if(num%i==0)
                return false;
        }
        return true;
    }

    public void dfs(int depth, String number, char[] numbers) {
        if(depth == numbers.length)
            return;

        for(int i=0; i<numbers.length; i++) {
            if(!visited[i]){
                primeCandidates.add(Integer.parseInt(number + numbers[i]));
                visited[i] = true;
                dfs(depth + 1, number + numbers[i], numbers);
                visited[i] = false;
            }
        }
    }

    public int solution(String numbers) {
        visited = new boolean[numbers.length()];
        dfs(0,"",numbers.toCharArray());
        int count = 0;
        for(int primeCandidate : primeCandidates) {
            if(isPrime(primeCandidate)) {
                count++;
            }
        }
        return count;
    }
}

File: 최윤경/소수 찾기.js
/*
[접근]

1. 숫자 조각을 하나씩 선택해서 만들 수 있는 모든 숫자 생성 → 완전탐색
2. (현재까지 만든 숫자, 사용한 숫자) 상태로 DFS 재귀
3. 현재 숫자 소수 여부 검사
4. 중복 제외 → Set으로 저장 후 개수 반환
*/

function solution(numbers) {
  const digits = numbers.split("");
  const result = new Set();
  const visited = Array(digits.length).fill(false);

  function isPrime(num) {
    if (num < 2) return false;

    for (let i = 2; i * i <= num; i++) {
      if (num % i === 0) return false;
    }

    return true;
  }

  function dfs(current) {
    if (current.length > 0) {
      const num = Number(current);

      if (isPrime(num)) {
        result.add(num);
      }
    }

    for (let i = 0; i < digits.length; i++) {
      if (visited[i]) continue;

      visited[i] = true;

      dfs(current + digits[i]);

      visited[i] = false;
    }
  }

  dfs("");

  return result.size;
}
